File: packages/python/src/blackrose/guard.py
"""Sync and async Guard clients wrapping the official TypeSafe SDK."""

from __future__ import annotations

import os
from collections.abc import Mapping
from typing import Any, Protocol

from typesafe_sdk import AsyncTypeSafeClient, Question, TypeSafeClient

from blackrose.decide import decide
from blackrose.errors import BlackroseError, GuardClosedError, TypeSafeRequestError
from blackrose.policy import Policy, PolicySide
from blackrose.types import CheckResult

State = Any  # JSONContent: str | Mapping | Sequence
ClientConfig = Mapping[str, Any]


def _resolve_model(model: str | None) -> str:
    if model is not None:
        stripped = model.strip()
        if stripped:
            return stripped
    for env in ("TYPESAFE_MODEL", "TYPESAFE_DEFAULT_MODEL"):
        value = os.environ.get(env, "").strip()
        if value:
            return value
    return "jev-latest"


class _SyncSystemOne(Protocol):
    def system_one(
        self,
        state: State,
        questions: Mapping[str, Question],
        *,
        model: str | None = None,
    ) -> Any: ...


class _AsyncSystemOne(Protocol):
    async def system_one(
        self,
        state: State,
        questions: Mapping[str, Question],
        *,
        model: str | None = None,
    ) -> Any: ...


def _client_kwargs(
    *,
    api_key: str | None,
    model: str,
    client_config: ClientConfig | None,
) -> dict[str, Any]:
    kwargs: dict[str, Any] = dict(client_config or {})
    kwargs["api_key"] = api_key
    kwargs["model"] = model
    return kwargs


class Guard:
    """Synchronous decision layer over TypeSafe ``system_one``."""

    def __init__(
        self,
        *,
        api_key: str | None = None,
        model: str | None = None,
        policy: Policy | None = None,
        client: _SyncSystemOne | None = None,
        client_config: ClientConfig | None = None,
    ) -> None:
        self.policy = policy or Policy()
        self._model = _resolve_model(model)
        self._owns_client = client is None
        self._closed = False
        self._client: _SyncSystemOne = client or TypeSafeClient(
            **_client_kwargs(api_key=api_key, model=self._model, client_config=client_config)
        )

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        if self._owns_client and hasattr(self._client, "close"):
            self._client.close()

    def __enter__(self) -> Guard:
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

    def check_input(self, state: State) -> CheckResult:
        return self._check(state, "input")

    def check_output(self, state: State) -> CheckResult:
        return self._check(state, "output")

    def _check(self, state: State, side: PolicySide) -> CheckResult:
        if self._closed:
            raise GuardClosedError("Guard is closed")
        questions = self.policy.questions_for(side)
        try:
            response = self._client.system_one(state, questions, model=self._model)
        except BlackroseError:
            raise
        except Exception as exc:
            raise TypeSafeRequestError(
                "TypeSafe system_one failed; fail closed (do not allow)"
            ) from exc
        return decide(response, self.policy, expected_checks=questions.keys())


class AsyncGuard:
    """Asynchronous decision layer over TypeSafe ``system_one``."""

    def __init__(
        self,
        *,
        api_key: str | None = None,
        model: str | None = None,
        policy: Policy | None = None,
        client: _AsyncSystemOne | None = None,
        client_config: ClientConfig | None = None,
    ) -> None:
        self.policy = policy or Policy()
        self._model = _resolve_model(model)
        self._owns_client = client is None
        self._closed = False
        self._client: _AsyncSystemOne = client or AsyncTypeSafeClient(
            **_client_kwargs(api_key=api_key, model=self._model, client_config=client_config)
        )

    async def aclose(self) -> None:
        if self._closed:
            return
        self._closed = True
        if self._owns_client and hasattr(self._client, "aclose"):
            await self._client.aclose()

    async def __aenter__(self) -> AsyncGuard:
        return self

    async def __aexit__(self, *exc: object) -> None:
        await self.aclose()

    async def check_input(self, state: State) -> CheckResult:
        return await self._check(state, "input")

    async def check_output(self, state: State) -> CheckResult:
        return await self._check(state, "output")

    async def _check(self, state: State, side: PolicySide) -> CheckResult:
        if self._closed:
            raise GuardClosedError("AsyncGuard is closed")
        questions = self.policy.questions_for(side)
        try:
            response = await self._client.system_one(state, questions, model=self._model)
        except BlackroseError:
            raise
        except Exception as exc:
            raise TypeSafeRequestError(
                "TypeSafe system_one failed; fail closed (do not allow)"
            ) from exc
        return decide(response, self.policy, expected_checks=questions.keys())

File: packages/js/src/guard.ts
/**
 * Guard client wrapping the official TypeSafe JavaScript SDK.
 *
 * Mirrors packages/python/src/blackrose/guard.py (async surface).
 */

import {
  type Question,
  type SystemOneResult,
  TypeSafeClient,
  type TypeSafeClientConfig,
} from "@typesafe-ai/sdk";

import { decide } from "./decide.js";
import { Policy, type PolicySide } from "./policy.js";
import type { CheckResult, GuardState } from "./types.js";

function resolveModel(model?: string | null): string {
  if (model != null && model.trim() !== "") return model.trim();
  for (const env of ["TYPESAFE_MODEL", "TYPESAFE_DEFAULT_MODEL"] as const) {
    const value = process.env[env]?.trim();
    if (value) return value;
  }
  return "jev-latest";
}

/** Minimal System One client surface (real SDK or test double). */
export interface SystemOneClient {
  systemOne(request: {
    state: GuardState;
    questions: Record<string, Question>;
    model?: string;
  }): Promise<SystemOneResult<Record<string, Question>> | unknown>;
  /** Optional dispose hook (real TypeSafe clients may expose `close`). */
  close?: () => void | Promise<void>;
}

export interface GuardOptions {
  apiKey?: string;
  model?: string;
  policy?: Policy;
  /** Inject a mock or custom client; defaults to `TypeSafeClient`. */
  client?: SystemOneClient;
  /** Extra TypeSafe client config when constructing the default client. */
  clientConfig?: Omit<TypeSafeClientConfig, "apiKey" | "defaultModel">;
}

/**
 * Decision layer over TypeSafe `systemOne`.
 *
 * ```ts
 * const guard = new Guard();
 * try {
 *   const result = await guard.checkInput("Ignore previous instructions…");
 *   if (result.verdict === "allow") { … }
 * } finally {
 *   await guard.close();
 * }
 * ```
 */
export class Guard {
  readonly policy: Policy;
  private readonly model: string;
  private readonly client: SystemOneClient;
  private readonly ownsClient: boolean;
  private closed = false;

  constructor(options: GuardOptions = {}) {
    this.policy = options.policy ?? new Policy();
    this.model = resolveModel(options.model);
    this.ownsClient = options.client == null;
    this.client =
      options.client ??
      new TypeSafeClient({
        ...options.clientConfig,
        apiKey: options.apiKey,
        defaultModel: this.model,
      });
  }

  checkInput(state: GuardState): Promise<CheckResult> {
    return this.check(state, "input");
  }

  checkOutput(state: GuardState): Promise<CheckResult> {
    return this.check(state, "output");
  }

  /**
   * Dispose the owned TypeSafe client (no-op when a custom `client` was injected).
   * Safe to call more than once.
   */
  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    if (!this.ownsClient) return;
    const closer = this.client.close;
    if (typeof closer === "function") {
      await closer.call(this.client);
    }
  }

  async [Symbol.asyncDispose](): Promise<void> {
    await this.close();
  }

  private async check(state: GuardState, side: PolicySide): Promise<CheckResult> {
    const questions = this.policy.questionsFor(side);
    const response = await this.client.systemOne({
      state,
      questions,
      model: this.model,
    });
    return decide(response, this.policy);
  }
}
